### PR TITLE
fix: Always dump shards

### LINKED ISSUE #1184: Always dump shards follow up and fixes
**Describe the bug**
> A clear and concise description of what the bug is.

Major problems:
1. flank downloads, parses and calculates shards twice (dump shards and test run)
2. flank run with only robo test (including sanity robo) is broken
```
version: local_snapshot
revision: 7ab394646734d47b6d48753a06c119de323426d0

Cannot dump shards for non instrumentation test, ensure test apk has been set.
```
3. Strange, duplicated shards output
```
version: local_snapshot
revision: 7ab394646734d47b6d48753a06c119de323426d0

 Smart Flank cache hit: 0% (0 / 11)
  Shard times: 1065s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s

Saved 6 shards to android_shards.json

[long flank config]

RunTests

 Smart Flank cache hit: 0% (0 / 11)
  Shard times: 1065s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s


 Smart Flank cache hit: 0% (0 / 9)
  Shard times: 1035s

  Uploading app-debug.apk .
  Uploading app-multiple-error-debug-androidTest.apk .  Uploading app-multiple-success-debug-androidTest.apk .  Uploading app-multiple-flaky-debug-androidTest.apk .
```

**To Reproduce**
> Steps to reproduce the behavior:

1. Run any test run

**Expected behavior**
> A clear and concise description of what you expected to happen.

1. We should reuse calculated shards
2. robo tests should be working
3. Output should not be duplicated

**Details (please complete the following information):**
> Have you tested on the latest Flank snapshot?
>
> Post the output of ` flank --version`.

```
version: local_snapshot
revision: 7ab394646734d47b6d48753a06c119de323426d0
```

### PR DESCRIPTION
**Pull request by adamfilipow92**
Fixes #1184 

1. Flank should always dump shards
2. Robo test should works
3. Obfuscation command should work on generate dump shards on the test run

## Checklist

- [X] Unit tested

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/cli/firebase/test/android/AndroidRunCommand.kt [hunk 1]
```diff
@@ -44,9 +44,9 @@ class AndroidRunCommand : CommonRunCommand(), Runnable {
             MockServer.start()
         }
 
-        val config = AndroidArgs.load(Paths.get(configPath), cli = this).validate()
+        val config = AndroidArgs.load(Paths.get(configPath), cli = this).validate().copy(obfuscateDumpShards = obfuscate)
```
Comment (adamfilipow92) [inline test_runner/src/main/kotlin/ftl/cli/firebase/test/android/AndroidRunCommand.kt:47]: If you have a better idea to pass the obfuscate parameter to newTestRun tell me please

**Comment by pawelpasterz** [comment]
@adamfilipow92 
I think this is working correctly (not finished testing though)
Just not sure if this is the proper approach. After some digging into code, I came up with an idea.
Let's extract shard calculation logic on `NewTestRun.kt` level. (If we implement dump and run tests logic as extension functions it could be event better IMO)
(pseudo code)
```kotlin
val (matrixMap, testShardChunks, ignoredTests) = cancelTestsOnTimeout(args.project) {
        shards = calculateShards
        shards.dump(dumpType)
        shards.runTests(args)
    }
```
1. no additional flag in config (`dumpType` can be type of enum, I don't like passing boolean as function parameter)
2. shards are reused
3. easier to write UT (IMO)

Some additional refactor would be required, example:
```kotlin
private suspend fun calculateShards(args: IArgs): shards {
    return when (args) {
        is AndroidArgs -> [android way to get shards]
        is IosArgs -> [ios way to get shards]
        else -> throw FlankGeneralError("err message")
    }
}
```
`runIosTests` is no-brainer to adjust, with `runAndroidTests` will be a bit more work.
We would need to implement something similar to `AndroidContext` but for iOS (or not?)

This is of course just a proposal!

My proposal is definitely more work but I think its cleaner and more flexible for future changes. But, as always, it's my subjective opinion.

If this feature is in much desire we can merge it and create an issue to refactor. @bootstraponline ?

File: test_runner/src/main/kotlin/ftl/run/platform/android/GetAndroidMatrixShards.kt [hunk 2]
```diff
@@ -6,12 +6,12 @@ import ftl.run.model.AndroidTestShards
 import ftl.run.model.InstrumentationTestContext
 import ftl.shard.testCases
 
-suspend fun AndroidArgs.getAndroidMatrixShards(): AndroidMatrixTestShards = this
-    .createAndroidTestContexts()
+suspend fun AndroidArgs.getAndroidMatrixShards(): AndroidMatrixTestShards =
+    createAndroidTestContexts()
```
[the remark this answers is not in the document]
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/run/platform/android/GetAndroidMatrixShards.kt]: Thanks, reverted

File: test_runner/src/main/kotlin/ftl/cli/firebase/test/android/AndroidRunCommand.kt [hunk 3]
```diff
@@ -44,9 +44,9 @@ class AndroidRunCommand : CommonRunCommand(), Runnable {
             MockServer.start()
         }
 
-        val config = AndroidArgs.load(Paths.get(configPath), cli = this).validate()
+        val config = AndroidArgs.load(Paths.get(configPath), cli = this).validate().copy()
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/cli/firebase/test/android/AndroidRunCommand.kt:47]:
```suggestion
        val config = AndroidArgs.load(Paths.get(configPath), cli = this).validate()
```

File: test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt [hunk 4]
```diff
@@ -19,7 +19,8 @@ data class AndroidArgs(
     val testRunnerClass: String?,
     val testTargets: List<String>,
     val additionalAppTestApks: List<AppTestPair>,
-    override val useLegacyJUnitResult: Boolean
+    override val useLegacyJUnitResult: Boolean,
+    val obfuscateDumpShards: Boolean
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:23]: this `val` is both in Android and iOS. Maybe move it to `CommonArgs`?

**Review by Sloox — APPROVED** [review]

**Review by jan-gogo — APPROVED** [review]

**Comment by piotradamczyk5** [comment]
@mergifyio refresh

**Pull request merged**